### PR TITLE
feat: add support for BYOID / STS

### PR DESCRIPTION
**Pull request by bshaffer**
@TimurSadykov @sai-sunder-s

### CODE REVIEW

File: tests/CredentialSource/FileSourceTest.php [hunk 1]
```diff
@@ -0,0 +1,89 @@
...
+use Google\Auth\CredentialSource\FileSource;
+use InvalidArgumentException;
+use PHPUnit\Framework\TestCase;
+use UnexpectedValueException;
+
+/**
+ * @group credentialsource
+ * @group credentialsource-file
+ */
+class FileSourceTest extends TestCase
```
Comment (aeitzman) [inline tests/CredentialSource/FileSourceTest.php:29]: Can you add tests for the happy path for both json and text format?

File: tests/CredentialSource/UrlSourceTest.php [hunk 2]
```diff
@@ -0,0 +1,144 @@
...
+/**
+ * @group credentialsource
+ * @group credentialsource-url
+ */
+class UrlSourceTest extends TestCase
+{
+    use ProphecyTrait;
+
+    /** @dataProvider provideFetchSubjectToken */
+    public function testFetchSubjectToken(
```
Comment (aeitzman) [inline tests/CredentialSource/UrlSourceTest.php:38]: Can you add a test for a successful call with JSON format?

File: tests/OAuth2Test.php [hunk 3]
```diff
@@ -1078,3 +1080,70 @@ public function testShouldReturnAValidIdToken()
...
+    }
+
+    public function testStsCredentialsRequestFull()
+    {
+        $credentialSource = $this->prophesize(CredentialSourceInterface::class);
+        $credentialSource->fetchSubjectToken(null)
+            ->shouldBeCalledOnce()
+            ->willReturn('xyz');
+        $stsMinimal = $this->stsMinimal + [
+            'subjectTokenFetcher' => $credentialSource->reveal(),
```
Comment (aeitzman) [inline tests/OAuth2Test.php:1126]: Can you also add a test (probably in the ExternalAccountCredentialsTests) thais oauth2/sts object is getting the correct values from the credential configuration when it is instantiated?

**Review by aeitzman — COMMENTED** [review]
Looks good, just a few nits and suggestions for test cases to add.

File: tests/CredentialSource/FileSourceTest.php (hunk 1 above)
Comment (aeitzman) [reply tests/CredentialSource/FileSourceTest.php:29]: Oh whoops, I did understand how these tests were running with the @dataProvider, I see now, my bad!

**Review by aeitzman — APPROVED** [review]

File: src/CredentialSource/UrlSource.php [hunk 4]
```diff
@@ -0,0 +1,94 @@
...
+
+    /**
+     * @param array<string, string|string[]> $headers Request headers
+     */
+    public function __construct(
+        string $url,
+        string $format = null,
+        string $subjectTokenFieldName = null,
+        array $headers = null
+    ) {
```
Comment (vishwarajanand) [inline src/CredentialSource/UrlSource.php:50]: document all params?

File: src/CredentialSource/FileSource.php [hunk 5]
```diff
@@ -0,0 +1,69 @@
...
+ */
+
+namespace Google\Auth\CredentialSource;
+
+use Google\Auth\ExternalAccountCredentialSourceInterface;
+use InvalidArgumentException;
+use UnexpectedValueException;
+
+/**
+ * Retrieve a token from a file.
```
[the remark this answers is not in the document]
Comment (bshaffer) [reply src/CredentialSource/FileSource.php:25]: I am going to remove `@internal` from UrlSource... I'm not quite sure how that got in there. Thanks for the callout!

**Pull request merged**